paths_get.py

```diff
@@ -0,0 +1,168 @@
+    return pd.DataFrame(columns, adata.obs_names)
+
+
+def get_obs_vector(adata: Union[AnnData, Raw], path: Union[str, AnnDataPath], gene_symbols: Union[str, AnnDataPath]):
+    path = normalize_adpath(path)
+    gene_symbols = None if gene_symbols is None else normalize_adpath(gene_symbols)
+    
+    if path[0] == "layer":  # X is here after normalizing
+        layer = adata.X if path[1] == "X" else adata.layers[path[1]]
+        idx = adata.var_names if gene_symbols is None else adata.var[gene_symbols]
+        # TODO: can layers also be data frames and have to be handled as obsm?
+        return layer[:, idx == path[2]]
+    if path[0] == "obs":
+        return adata.obs[path[1]]
+    assert not isinstance(adata, Raw)
+    if path[0] == "obsm":
+        m = adata.obsm[path[1]]
+        return m[path[2]] if isinstance(m, pd.DataFrame) else m[:, path[2]]
+    if path[0] == "obsp":
+        raise NotImplementedError("No good idea how to specify this. 'obsp/neighbors/CellName/1 vs .../2?'")
+    if path[0] == "raw":
+        return get_obs_vector(adata.raw, path[1:], gene_symbols)
+    else:
+        raise ValueError(f"Unknown AnnData attribute {path[0]!r} in path {path!r}")
+
+
+SHORT_CODES = dict(
+    X=["layer", "X"],
+    l=["layer"],
+    o=["obs"],
+    m=["obsm"],
+    p=["obsp"],
+    rX=["raw", "X"],
+    ro=["raw", "obs"],
+)
+
+
+def normalize_adpath(path: Union[str, AnnDataPath]) -> AnnDataPath:
+    # path is shorthand: "obs/Foo" and "o/Foo"
+    if not isinstance(path, str):
+        return path
+    path = path.split("/")
+    path[:1] = SHORT_CODES.get(path[0], [path[0]])
+    if path[0] in {"obsm", "obsp"}:
+        # TODO: how to normalize
+        try:
+            path[-1] = int(path[-1])
+        except ValueError:
+            pass
+    return tuple(path)
+
+
+def paths_to_names(paths: Sequence[AnnDataPath], length: int = 1) -> Dict[str, AnnDataPath]:
+    names = {}
+    dupes = {}
+    for path, name in zip(paths, (path_name(p, length) for p in paths)):
+        dupes.setdefault(name, []).append(path)
+    for name, paths_dup in dupes.items():
+        if len(paths_dup) == 1:
+            names[name] = paths_dup[0]
+        elif any(len(p) > length for p in paths_dup):
+            names.update(paths_to_names(paths_dup, length + 1))
+        else:
+            raise ValueError(f"Not sure how {name} can be extended for {paths_dup}")
+    return names
+
+
+def path_name(path: AnnDataPath, length: int = 1) -> str:
+    if length == 1 and path[0] == "obsp":
+        return path[-2]  # just key
+    if length in {1, 2} and path[0] == "obsm":
+        if isinstance(path[-1], int):
+            return f"{path[-2]}{path[-1]+1}"  # X_pca1
+        else:  # normal
+            return path[-1] if length == 1 else f"{path[-2]}-{path[-1]}"
+    if length <= 2:
+        return "-".join(path[-length:])
+    return f"{path[:-2]}{path_name(path, length=2)}"
+    
+    
+
+
+def old():
```

Comment:
Oh, I don't think this will be needed anymore after https://github.com/theislab/anndata/pull/340. Probably just refer to the version on master. 

How ugly this version turned out is why I hadn't made a PR for https://github.com/theislab/scanpy/pull/1109 until we were talking about it.

Reply:
I just kept it around in case I had to move code over before I delete it.

I’ll look into your PR.